### PR TITLE
[RISCV] Match vcompress during shuffle lowering

### PR DESCRIPTION
**Pull request by preames**
This change matches a subset of vcompress patterns during shuffle lowering. The subset implemented requires a contiguous prefix of demanded elements followed by undefs.  This subset was chosen for two reasons: 1) which elements to spurious demand is a non-obvious problem, and 2) my first several attempts at implementing the general case were buggy.  I decided to go with the simple case to start with.

vcompress scales better with LMUL than a general vrgather, and at least the SpaceMit X60, has higher throughput even at m1.  It also has the advantage of requiring smaller vector constants at one bit per element as opposed to vrgather which is a minimum of 8 bits per element.  The downside to using vcompress is that we can't fold a vselect into it, as there is no masked vcompress variant.

For reference, here are the relevant throughputs from camel-cdr's data table on BP3 (X60):
  vrgather.vv v8,v16,v24    4.0  16.0  64.0  256.0
  vcompress.vm v8,v16,v24   3.0  10.0  36.0  136.
  vmerge.vvm v8,v16,v24,v0  2.0  4.0   8.0   16.0

The largest concern with the extra vmerge is that we locally increase register pressure.  If we do have masking, we also have a passthru, without the ability to fold that into the vcompress, we need to keep it alive a bit longer.  This can hurt at e.g. m8 where we have very few architectural registers.  As compared with the vrgather.vv sequence, this is only one additional m1 VREG - since we no longer need the index vector. It compares slightly worse against vrgatherie16.vv which can use index vectors smaller than other operands.  Note that we could potentially fold the vmerge if only tail elements are being preserved; I haven't investigated this.

It is unfortunately hard given our current lowering structure to know if we're emitting a shuffle where masking will follow.  Thankfully, it doesn't seem to show up much in practice, so I think we can probably ignore it.

This patch only handles single source compress idioms at the moment. This is an effort to avoid interacting with other patches on review for changing how we canonicalize length changing shuffles.

### CODE REVIEW

**Review by lukel97 — APPROVED** [review]
LGTM. I'm not concerned about the lack of masking, I think the benefit of being able to sometimes materialize the mask as a single li + vmv.s.x is enough to outweigh it.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -5155,6 +5155,28 @@ static SDValue lowerShuffleViaVRegSplitting(ShuffleVectorSDNode *SVN,
...
+  int Last = -1;
+  bool SawUndef = false;
+  for (int i = 0; i < Mask.size(); i++) {
+    if (Mask[i] == -1) {
+      SawUndef = true;
+      continue;
+    }
+    if (SawUndef)
+      return false;
+    if (i > (unsigned)Mask[i])
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5171]: Why do we need a cast to unsigned? `i` and `Mask[i]` are both `int`.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -5372,6 +5394,25 @@ static SDValue lowerVECTOR_SHUFFLE(SDValue Op, SelectionDAG &DAG,
     if (SDValue V = lowerVECTOR_SHUFFLEAsRotate(SVN, DAG, Subtarget))
       return V;
 
+    // Can we generate a vcompress instead of a vrgather?  These scale better
+    // at high LMUL, at the cost of not being able to fold a following select
+    // into them.  The mask constants are also smaller than the index vector
+    // constants, and thus easier to materialize.
+    if (isCompressMask(Mask)) {
+      SmallVector<SDValue> MaskVals;
+      MaskVals.resize(NumElts, DAG.getConstant(false, DL, XLenVT));
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5403]: Use the SmallVector constructor to do this?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**